### ISSUE TITLE
exec auth provider expiration causes problems

### ISSUE THREAD

**Issue by nicks** [windmilleng/tilt#2702]
Repro steps:
1) Check out any Tilt project and sync to the latest version of https://github.com/windmilleng/k8s-rbac-testing, our library of auth testing scripts
2) Create a KinD cluster (`kind create cluster`)
3) Create a namespace guarded by an exec auth provider: (`./k8s-rbac-testing/create-restricted-namespace.sh -e n1`)
4) Our script will create two tempfiles. One contains the auth credentials. The other contains an access log for the auth credentials. It will look like this:

```
A credential script has been created. Rotate the keys with:
./rotate-exec-credential-script.sh n5 /tmp/tmp.t5DZcipNY6

Get the access log with:
cat /tmp/tmp.t5DZcipNY6.log
```

Go into that first tempfile, and add an expirationTimestamp for sometime in the past

```
{
  "kind": "ExecCredential",
  "apiVersion": "client.authentication.k8s.io/v1beta1",
  "status": {
    "expirationTimestamp": "2019-12-20T16:12:02Z",
    ...
}
```

5) Run `tilt up` with the KUBECONFIG created in step 3. Tail the access log reported in step (3)

Expected result:.
Tilt should somehow report that your token is expired, and incrementally back off

Actual result:
Tilt keeps calling the exec auth provider dozens of times per minute. The token works ok.

If you delete the token (`./k8s-rbac-testing/delete-exec-credential-token.sh n1`) while Tilt is up, even weirder things happen. 

The informers don't initially notice that they've lost access to the cluster. When you change something, Tilt will notice that it can't make changes to the cluster and report errors appropriately. Each push accesses the exec auth provider twice.

About 5 minutes later the informers blow up and start fetching 5-10x a second with the error "Failed to list XXX resource" and accessing the exec provider that frequently

**Comment by nicks** [comment]
oh, wait, i'm wrong! the informers do notice. they blow up and eventually start fetching 5-10x a second with the error "Failed to list XXX resource" and accessing the exec provider that often

**Comment by nicks** [comment]
updated the above comment with more details on how the informers behave. We may need to file a bug upstream against client-go

**Comment by nicks** [comment]
also, some historical context: @jazzdan has talked to sig-api-machinery in the past with the problems with how client-go informers handle and report errors, but there hasn't been much movement towards fixing the problem. See https://github.com/kubernetes/client-go/issues/155 for additional discussion.

It might be worth forking client-go, coming up with a solution that works for us, then proposing it upstream

**Comment by nicks** [comment]
Captured a stack trace from the Informers

```
REFRESH CREDS: goroutine 20594 [running]:
runtime/debug.Stack(0x209f5a0, 0xc000981ad0, 0xc0001c28f0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
k8s.io/client-go/plugin/pkg/client/auth/exec.(*Authenticator).refreshCredsLocked(0xc0004f6dd0, 0x0, 0x0, 0x0)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:285 +0x42
k8s.io/client-go/plugin/pkg/client/auth/exec.(*Authenticator).getCreds(0xc0004f6dd0, 0x0, 0x0, 0x0)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:260 +0xb5
k8s.io/client-go/plugin/pkg/client/auth/exec.(*roundTripper).RoundTrip(0xc00067a3a0, 0xc000104400, 0x20f6d18, 0xa, 0xc00041fc20)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:214 +0x79
k8s.io/client-go/transport.(*userAgentRoundTripper).RoundTrip(0xc00067a3e0, 0xc000104300, 0xc00067a3e0, 0x0, 0x0)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/transport/round_trippers.go:159 +0x1c0
net/http.send(0xc000104300, 0x2454cc0, 0xc00067a3e0, 0x0, 0x0, 0x0, 0xc0003120a8, 0xc000df8080, 0x1, 0x0)
	/usr/local/go/src/net/http/client.go:250 +0x443
net/http.(*Client).send(0xc0002a2330, 0xc000104300, 0x0, 0x0, 0x0, 0xc0003120a8, 0x0, 0x1, 0xc000b24120)
	/usr/local/go/src/net/http/client.go:174 +0xfa
net/http.(*Client).do(0xc0002a2330, 0xc000104300, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/client.go:641 +0x3ce
net/http.(*Client).Do(0xc0002a2330, 0xc000104300, 0x0, 0x3, 0xc000cbe000)
	/usr/local/go/src/net/http/client.go:509 +0x35
k8s.io/client-go/rest.(*Request).request(0xc000994180, 0xc0004d1680, 0x0, 0x0)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/rest/request.go:754 +0x367
k8s.io/client-go/rest.(*Request).Do(0xc000994180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/rest/request.go:828 +0xd8
k8s.io/client-go/kubernetes/typed/core/v1.(*services).List(0xc000b06020, 0x0, 0x0, 0x0, 0x0, 0xc0007a2000, 0x21, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/kubernetes/typed/core/v1/service.go:90 +0x1a1
k8s.io/client-go/informers/core/v1.NewFilteredServiceInformer.func1(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x20ebcbf, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/informers/core/v1/service.go:64 +0x194
k8s.io/client-go/tools/pager.SimplePageFunc.func1(0x24ae260, 0xc000056168, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/pager/pager.go:40 +0x64
k8s.io/client-go/tools/pager.(*ListPager).List(0xc0004d1b98, 0x24ae260, 0xc000056168, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/pager/pager.go:88 +0x12d
k8s.io/client-go/tools/cache.(*ListWatch).List(0xc00067b840, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/cache/listwatch.go:106 +0x16e
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch.func1.1.2(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x20ebcbf, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/cache/reflector.go:182 +0x72
k8s.io/client-go/tools/pager.SimplePageFunc.func1(0x24ae260, 0xc000056160, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/pager/pager.go:40 +0x64
k8s.io/client-go/tools/pager.(*ListPager).List(0xc0004d1f10, 0x24ae260, 0xc000056160, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/pager/pager.go:88 +0x12d
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch.func1.1(0xc0007aa180, 0xc0006421e0, 0xc00057e380, 0xc0008f4020, 0xc0008f4030, 0xc00005c1e0)
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/cache/reflector.go:188 +0x1a4
created by k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch.func1
	/home/nick/golang/src/github.com/windmilleng/tilt/vendor/k8s.io/client-go/tools/cache/reflector.go:173 +0x2af
```

**Comment by nicks** [comment]
also this snapshot captures what happens when the first kubectl apply fails:

https://cloud.tilt.dev/snapshot/AeT3-uELeYzX34rxtcc=

there's some sort of snapshots bug that's preventing me from posting the error when the informers fail, but that's a separate problem

**Comment by nicks** [comment]
ooh, here's a better snapshot i was able to capture:
https://cloud.tilt.dev/snapshot/AdSc--ELaQUbfMmsMBU=

**Comment by nicks** [comment]
ok, here's the problem code:
https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L162

Run() calls ListAndWatch() every second, with no way to change the period, or insert any sort of backoff.

I think the next steps are:
1) Stop using Run(). Invoke ListAndWatch() ourselves, with our own incremental backoff and error handling
2) File a bug upstream with client-go. At the very least, it's bizarre that the backoff period isn't configurable

**Comment by nicks** [comment]
this should be fixed in v0.11.1
